compaction: update maintenance sstable set on scrub compaction completion

lkshminarayanan opened the pull request:
Scrub compaction can pick up input sstables from maintenance sstable set
but on compaction completion, it doesn't update the maintenance set
leaving the original sstable in set after it has been scrubbed. To fix
this, on compaction completion has to update the maintenance sstable if
the input originated from there. This PR solves the issue by updating the
correct sstable_sets on compaction completion.

Fixes #20030

This issue has existed since the introduction of main and maintenance sstable sets into scrub compaction. It would be good to have the fix backported to versions 6.1 and 6.2.

lkshminarayanan commented:
> > The PR achieves this by introducing a bool flag
> > using which the origin of the sstable is tracked throughout the scrub
> > compaction process.
> 
> I haven't reviewed the code yet, but right away a bool flag sounds too restrictive. We have many different sstable directories, how can we handle selecting the correct set to remove the sstable from by just a boolean?

The flag is used to identify which sstable_set the sstable belongs to - either main_sstable_set or maintenance_sstable_set. I 'll update the cover letter.

denesb commented:
> > > The PR achieves this by introducing a bool flag
> > > using which the origin of the sstable is tracked throughout the scrub
> > > compaction process.
> > 
> > 
> > I haven't reviewed the code yet, but right away a bool flag sounds too restrictive. We have many different sstable directories, how can we handle selecting the correct set to remove the sstable from by just a boolean?
> 
> The flag is used to identify which sstable_set the sstable belongs to - either main_sstable_set or maintenance_sstable_set. I'll update the cover letter.

But can we not scrub sstables in other directories?

lkshminarayanan commented:
> > > > The PR achieves this by introducing a bool flag
> > > > using which the origin of the sstable is tracked throughout the scrub
> > > > compaction process.
> > > 
> > > 
> > > I haven't reviewed the code yet, but right away a bool flag sounds too restrictive. We have many different sstable directories, how can we handle selecting the correct set to remove the sstable from by just a boolean?
> > 
> > 
> > The flag is used to identify which sstable_set the sstable belongs to - either main_sstable_set or maintenance_sstable_set. I'll update the cover letter.
> 
> But can we not scrub sstables in other directories?

The regular scrubbing picks up only sstables from these two sets - relevant code [here](https://github.com/scylladb/scylladb/blob/1d0c6aa26f105056aab001022a0f9487850af16b/compaction/compaction_manager.cc#L2153).

There is a validate mode that works on quarantined sstables but this PR will not affect that path.

lkshminarayanan commented:
> > > > > > The PR achieves this by introducing a bool flag
> > > > > > using which the origin of the sstable is tracked throughout the scrub
> > > > > > compaction process.
> > > > > 
> > > > > 
> > > > > I haven't reviewed the code yet, but right away a bool flag sounds too restrictive. We have many different sstable directories, how can we handle selecting the correct set to remove the sstable from by just a boolean?
> > > > 
> > > > 
> > > > The flag is used to identify which sstable_set the sstable belongs to - either main_sstable_set or maintenance_sstable_set. I'll update the cover letter.
> > > 
> > > 
> > > But can we not scrub sstables in other directories?
> > 
> > 
> > The regular scrubbing picks up only sstables from these two sets - relevant code [here](https://github.com/scylladb/scylladb/blob/1d0c6aa26f105056aab001022a0f9487850af16b/compaction/compaction_manager.cc#L2153).
> > There is a validate mode that works on quarantined sstables but this PR will not affect that path.
> 
> Maybe we can merge update_sstable_lists_on_off_strategy_completion() into update_main_sstable_list_on_compaction_completion(), so we can get rid of the flag (even existing offstrategy one).

Sounds good. So, we rebuild both main and maintenance sets at the end of every compaction regardless of they are offstrategy or not. Is that okay?

raphaelsc commented:
> > > > > > > The PR achieves this by introducing a bool flag
> > > > > > > using which the origin of the sstable is tracked throughout the scrub
> > > > > > > compaction process.
> > > > > > 
> > > > > > 
> > > > > > I haven't reviewed the code yet, but right away a bool flag sounds too restrictive. We have many different sstable directories, how can we handle selecting the correct set to remove the sstable from by just a boolean?
> > > > > 
> > > > > 
> > > > > The flag is used to identify which sstable_set the sstable belongs to - either main_sstable_set or maintenance_sstable_set. I'll update the cover letter.
> > > > 
> > > > 
> > > > But can we not scrub sstables in other directories?
> > > 
> > > 
> > > The regular scrubbing picks up only sstables from these two sets - relevant code [here](https://github.com/scylladb/scylladb/blob/1d0c6aa26f105056aab001022a0f9487850af16b/compaction/compaction_manager.cc#L2153).
> > > There is a validate mode that works on quarantined sstables but this PR will not affect that path.
> > 
> > 
> > Maybe we can merge update_sstable_lists_on_off_strategy_completion() into update_main_sstable_list_on_compaction_completion(), so we can get rid of the flag (even existing offstrategy one).
> 
> Sounds good. So, we rebuild both main and maintenance sets at the end of every compaction regardless of they are offstrategy or not. Is that okay?

That's exactly what I thought, removing the input sstables from both (set that doesn't have it won't be affected by attempt to remove sstable that exist in the other set), and adding to main, since it's one-directional (always output is added to main set).

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13170/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13170/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/sstable_compaction_test
:white_check_mark: - [Docker Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-docker/222/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/4765/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/12788/)
:x: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/1689/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13170/)
#### Build Details:
- Duration: 8 hr 29 min
- Builder: i-0d55acab174bddcd0 (m5ad.8xlarge)

lkshminarayanan commented:
Changelog V2 :
 - Existing `update_main_sstable_list_on_compaction_completion` method has been repurposed to update the maintenance sstable sets as well on compaction completion.
 - The `update_main_sstable_list_on_compaction_completion` is now renamed to `update_sstable_sets_on_compaction_completion`
 - replaced `update_sstable_lists_on_off_strategy_completion` with the new `update_sstable_sets_on_compaction_completion`

lkshminarayanan commented:
> > > > > > > > The PR achieves this by introducing a bool flag
> > > > > > > > using which the origin of the sstable is tracked throughout the scrub
> > > > > > > > compaction process.
> > > > > > > 
> > > > > > > 
> > > > > > > I haven't reviewed the code yet, but right away a bool flag sounds too restrictive. We have many different sstable directories, how can we handle selecting the correct set to remove the sstable from by just a boolean?
> > > > > > 
> > > > > > 
> > > > > > The flag is used to identify which sstable_set the sstable belongs to - either main_sstable_set or maintenance_sstable_set. I'll update the cover letter.
> > > > > 
> > > > > 
> > > > > But can we not scrub sstables in other directories?
> > > > 
> > > > 
> > > > The regular scrubbing picks up only sstables from these two sets - relevant code [here](https://github.com/scylladb/scylladb/blob/1d0c6aa26f105056aab001022a0f9487850af16b/compaction/compaction_manager.cc#L2153).
> > > > There is a validate mode that works on quarantined sstables but this PR will not affect that path.
> > > 
> > > 
> > > Maybe we can merge update_sstable_lists_on_off_strategy_completion() into update_main_sstable_list_on_compaction_completion(), so we can get rid of the flag (even existing offstrategy one).
> > 
> > 
> > Sounds good. So, we rebuild both main and maintenance sets at the end of every compaction regardless of they are offstrategy or not. Is that okay?
> 
> That's exactly what I thought, removing the input sstables from both (set that doesn't have it won't be affected by attempt to remove sstable that exist in the other set), and adding to main, since it's one-directional (always output is added to main set).

Done. cc @denesb

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13208/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13208/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/sstable_compaction_test
:white_check_mark: - [Docker Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-docker/255/)
:x: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/1730/)
:x: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/12827/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13208/)
:x: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/4821/)
#### Failed Tests (362/36624):
- [test_no_cleanup_when_unnecessary.debug.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13208/testReport/junit/%28root%29/test_automatic_cleanup/Tests___Unit_Tests___test_no_cleanup_when_unnecessary_debug_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_no_cleanup_when_unnecessary.debug.2)
- [test_restore_snapshot_using_old_token_ownership](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13208/testReport/junit/backup_restore_tests/TestBackupRestore/Tests___dtest___test_restore_snapshot_using_old_token_ownership) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_restore_snapshot_using_old_token_ownership)
- [test_restore_snapshot_using_old_token_ownership](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13208/testReport/junit/backup_restore_tests/TestBackupRestore/Tests___dtest___test_restore_snapshot_using_old_token_ownership_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_restore_snapshot_using_old_token_ownership)
- [test_restore_snapshot_using_old_token_ownership](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13208/testReport/junit/backup_restore_tests/TestBackupRestore/Tests___dtest_with_topology_changes___test_restore_snapshot_using_old_token_ownership) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_restore_snapshot_using_old_token_ownership)
- [test_restore_snapshot_using_old_token_ownership](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13208/testReport/junit/backup_restore_tests/TestBackupRestore/Tests___dtest_with_topology_changes___test_restore_snapshot_using_old_token_ownership_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_restore_snapshot_using_old_token_ownership)
- [test_manual_bootstrap](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13208/testReport/junit/bootstrap_test/TestBootstrap/Tests___dtest___test_manual_bootstrap) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_manual_bootstrap)
- [test_manual_bootstrap](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13208/testReport/junit/bootstrap_test/TestBootstrap/Tests___dtest_with_topology_changes___test_manual_bootstrap) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_manual_bootstrap)
- [test_off_strategy_during_bootstrap](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13208/testReport/junit/bootstrap_test/TestBootstrap/Tests___dtest___test_off_strategy_during_bootstrap) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_off_strategy_during_bootstrap)
- [test_off_strategy_during_bootstrap](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13208/testReport/junit/bootstrap_test/TestBootstrap/Tests___dtest_with_topology_changes___test_off_strategy_during_bootstrap) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_off_strategy_during_bootstrap)
- [test_cluster_cleanup](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13208/testReport/junit/cleanup_test/TestCleanup/Tests___dtest___test_cluster_cleanup) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_cluster_cleanup)
#### Build Details:
- Duration: 8 hr 9 min
- Builder: spider8.cloudius-systems.com

raphaelsc commented on replica/table.cc:
```diff
@@ -1915,11 +1916,25 @@ compaction_group::update_sstable_sets_on_compaction_completion(sstables::compact
+                }
             }
             // FIXME: the following is not exception safe
             _t.refresh_compound_sstable_set();
```
only old sstables from main should be removed from the backlog tracker below. maybe build_new_list() interface can be adjusted to return which sstables were actually removed from the set (using pair or struct), and then you can use that info to simply condition to trigger maintenance set rebuild, and adjust backlog tracker appropriately.

lkshminarayanan replied:
Done.

lkshminarayanan commented:
Changelog V3:
 - updated `build_new_list()` to return removed sstables as well and use that to update the main/maintenance sstable sets.

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13271/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13271/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/sstable_compaction_test
:white_check_mark: - [Docker Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-docker/304/)
:x: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/1799/)
:x: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/12904/)
:x: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/4892/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13271/)
#### Failed Tests (10/36609):
- [test_delete_tombstone_gc_node_down[TimeWindowCompactionStrategy]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13271/testReport/junit/compaction_test/TestCompaction/Tests___dtest_with_tablets___test_delete_tombstone_gc_node_down_TimeWindowCompactionStrategy_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_delete_tombstone_gc_node_down[TimeWindowCompactionStrategy])
- [test_delete_tombstone_gc_node_down[LeveledCompactionStrategy]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13271/testReport/junit/compaction_test/TestCompaction/Tests___dtest_with_tablets___test_delete_tombstone_gc_node_down_LeveledCompactionStrategy_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_delete_tombstone_gc_node_down[LeveledCompactionStrategy])
- [test_delete_tombstone_gc_node_down[SizeTieredCompactionStrategy]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13271/testReport/junit/compaction_test/TestCompaction/Tests___dtest_with_tablets___test_delete_tombstone_gc_node_down_SizeTieredCompactionStrategy_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_delete_tombstone_gc_node_down[SizeTieredCompactionStrategy])
- [test_simple_add_new_node_while_query_info[1]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13271/testReport/junit/update_cluster_layout_tests/TestUpdateClusterLayout/Tests___dtest___test_simple_add_new_node_while_query_info_1_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_simple_add_new_node_while_query_info[1])
- [test_delete_tombstone_gc_node_down[SizeTieredCompactionStrategy]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13271/testReport/junit/compaction_test/TestCompaction/Tests___dtest___test_delete_tombstone_gc_node_down_SizeTieredCompactionStrategy_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_delete_tombstone_gc_node_down[SizeTieredCompactionStrategy])
- [test_delete_tombstone_gc_node_down[LeveledCompactionStrategy]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13271/testReport/junit/compaction_test/TestCompaction/Tests___dtest___test_delete_tombstone_gc_node_down_LeveledCompactionStrategy_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_delete_tombstone_gc_node_down[LeveledCompactionStrategy])
- [test_delete_tombstone_gc_node_down[TimeWindowCompactionStrategy]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13271/testReport/junit/compaction_test/TestCompaction/Tests___dtest___test_delete_tombstone_gc_node_down_TimeWindowCompactionStrategy_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_delete_tombstone_gc_node_down[TimeWindowCompactionStrategy])
- [test_delete_tombstone_gc_node_down[SizeTieredCompactionStrategy]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13271/testReport/junit/compaction_test/TestCompaction/Tests___dtest_with_topology_changes___test_delete_tombstone_gc_node_down_SizeTieredCompactionStrategy_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_delete_tombstone_gc_node_down[SizeTieredCompactionStrategy])
- [test_delete_tombstone_gc_node_down[LeveledCompactionStrategy]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13271/testReport/junit/compaction_test/TestCompaction/Tests___dtest_with_topology_changes___test_delete_tombstone_gc_node_down_LeveledCompactionStrategy_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_delete_tombstone_gc_node_down[LeveledCompactionStrategy])
- [test_delete_tombstone_gc_node_down[TimeWindowCompactionStrategy]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13271/testReport/junit/compaction_test/TestCompaction/Tests___dtest_with_topology_changes___test_delete_tombstone_gc_node_down_TimeWindowCompactionStrategy_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_delete_tombstone_gc_node_down[TimeWindowCompactionStrategy])
#### Build Details:
- Duration: 8 hr 43 min
- Builder: i-0797c39d6bf58af00 (m5ad.8xlarge)

raphaelsc commented:
> #### 🔴 CI State: FAILURE
> ✅ - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13271/) ✅ - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13271/) **The following new/updated tests ran 100 times for each mode:** 🔹 boost/sstable_compaction_test ✅ - [Docker Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-docker/304/) ❌ - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/1799/) ❌ - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/12904/) ❌ - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/4892/) ✅ - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13271/)
> 
> #### Failed Tests (10/36609):
> * [test_delete_tombstone_gc_node_down[TimeWindowCompactionStrategy]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13271/testReport/junit/compaction_test/TestCompaction/Tests___dtest_with_tablets___test_delete_tombstone_gc_node_down_TimeWindowCompactionStrategy_) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_delete_tombstone_gc_node_down%5BTimeWindowCompactionStrategy%5D)
> * [test_delete_tombstone_gc_node_down[LeveledCompactionStrategy]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13271/testReport/junit/compaction_test/TestCompaction/Tests___dtest_with_tablets___test_delete_tombstone_gc_node_down_LeveledCompactionStrategy_) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_delete_tombstone_gc_node_down%5BLeveledCompactionStrategy%5D)
> * [test_delete_tombstone_gc_node_down[SizeTieredCompactionStrategy]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13271/testReport/junit/compaction_test/TestCompaction/Tests___dtest_with_tablets___test_delete_tombstone_gc_node_down_SizeTieredCompactionStrategy_) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_delete_tombstone_gc_node_down%5BSizeTieredCompactionStrategy%5D)
> * [test_simple_add_new_node_while_query_info[1]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13271/testReport/junit/update_cluster_layout_tests/TestUpdateClusterLayout/Tests___dtest___test_simple_add_new_node_while_query_info_1_) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_simple_add_new_node_while_query_info%5B1%5D)
> * [test_delete_tombstone_gc_node_down[SizeTieredCompactionStrategy]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13271/testReport/junit/compaction_test/TestCompaction/Tests___dtest___test_delete_tombstone_gc_node_down_SizeTieredCompactionStrategy_) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_delete_tombstone_gc_node_down%5BSizeTieredCompactionStrategy%5D)
> * [test_delete_tombstone_gc_node_down[LeveledCompactionStrategy]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13271/testReport/junit/compaction_test/TestCompaction/Tests___dtest___test_delete_tombstone_gc_node_down_LeveledCompactionStrategy_) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_delete_tombstone_gc_node_down%5BLeveledCompactionStrategy%5D)
> * [test_delete_tombstone_gc_node_down[TimeWindowCompactionStrategy]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13271/testReport/junit/compaction_test/TestCompaction/Tests___dtest___test_delete_tombstone_gc_node_down_TimeWindowCompactionStrategy_) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_delete_tombstone_gc_node_down%5BTimeWindowCompactionStrategy%5D)
> * [test_delete_tombstone_gc_node_down[SizeTieredCompactionStrategy]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13271/testReport/junit/compaction_test/TestCompaction/Tests___dtest_with_topology_changes___test_delete_tombstone_gc_node_down_SizeTieredCompactionStrategy_) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_delete_tombstone_gc_node_down%5BSizeTieredCompactionStrategy%5D)
> * [test_delete_tombstone_gc_node_down[LeveledCompactionStrategy]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13271/testReport/junit/compaction_test/TestCompaction/Tests___dtest_with_topology_changes___test_delete_tombstone_gc_node_down_LeveledCompactionStrategy_) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_delete_tombstone_gc_node_down%5BLeveledCompactionStrategy%5D)
> * [test_delete_tombstone_gc_node_down[TimeWindowCompactionStrategy]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13271/testReport/junit/compaction_test/TestCompaction/Tests___dtest_with_topology_changes___test_delete_tombstone_gc_node_down_TimeWindowCompactionStrategy_) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_delete_tombstone_gc_node_down%5BTimeWindowCompactionStrategy%5D)
> 
> #### Build Details:
> * Duration: 8 hr 43 min
> * Builder: i-0797c39d6bf58af00 (m5ad.8xlarge)

@lkshminarayanan any idea on the failure?

avikivity commented:
> Scrub compaction can pick up input sstables from maintenance sstable set but on compaction completion, it doesn't update the maintenance set leaving the original sstable in set after it has been scrubbed. To fix this, on compaction completion has to update the maintenance sstable if the input originated from there. This PR solves the issue by updating the correct sstable_sets on compaction completion.
> 
> Fixes #20030
> 
> This is a bug fix and should be backported to older versions.

How does scrub change sstables? Isn't it a read-only process?

raphaelsc commented:
> > Scrub compaction can pick up input sstables from maintenance sstable set but on compaction completion, it doesn't update the maintenance set leaving the original sstable in set after it has been scrubbed. To fix this, on compaction completion has to update the maintenance sstable if the input originated from there. This PR solves the issue by updating the correct sstable_sets on compaction completion.
> > Fixes #20030
> > This is a bug fix and should be backported to older versions.
> 
> How does scrub change sstables? Isn't it a read-only process?

There's a scrub mode that tries to repair logical damage, like reordered keys IIRC. I am not a fan of it, repair is better to fix any kind of corruption. And I am also not a fan of putting the repaired sstable online right away. But anyway, this is orthogonal to this PR.

denesb commented:
> > Scrub compaction can pick up input sstables from maintenance sstable set but on compaction completion, it doesn't update the maintenance set leaving the original sstable in set after it has been scrubbed. To fix this, on compaction completion has to update the maintenance sstable if the input originated from there. This PR solves the issue by updating the correct sstable_sets on compaction completion.
> > Fixes #20030
> > This is a bug fix and should be backported to older versions.
> 
> How does scrub change sstables? Isn't it a read-only process?

Actually, all but `VALIDATE` scrub mode, do modify the sstable -- they rewrite it into a new output sstable and discard the input sstables, just like all other compaction type. The `VALIDATE` is an exception. In this mode, the only change is that sstables may be moved to the quarantine directory (IIRC this can be disabled via the API).

lkshminarayanan commented:
> > #### 🔴 CI State: FAILURE
> > ✅ - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13271/) ✅ - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13271/) **The following new/updated tests ran 100 times for each mode:** 🔹 boost/sstable_compaction_test ✅ - [Docker Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-docker/304/) ❌ - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/1799/) ❌ - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/12904/) ❌ - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/4892/) ✅ - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13271/)
> > #### Failed Tests (10/36609):
> > 
> > * [test_delete_tombstone_gc_node_down[TimeWindowCompactionStrategy]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13271/testReport/junit/compaction_test/TestCompaction/Tests___dtest_with_tablets___test_delete_tombstone_gc_node_down_TimeWindowCompactionStrategy_) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_delete_tombstone_gc_node_down%5BTimeWindowCompactionStrategy%5D)
> > * [test_delete_tombstone_gc_node_down[LeveledCompactionStrategy]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13271/testReport/junit/compaction_test/TestCompaction/Tests___dtest_with_tablets___test_delete_tombstone_gc_node_down_LeveledCompactionStrategy_) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_delete_tombstone_gc_node_down%5BLeveledCompactionStrategy%5D)
> > * [test_delete_tombstone_gc_node_down[SizeTieredCompactionStrategy]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13271/testReport/junit/compaction_test/TestCompaction/Tests___dtest_with_tablets___test_delete_tombstone_gc_node_down_SizeTieredCompactionStrategy_) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_delete_tombstone_gc_node_down%5BSizeTieredCompactionStrategy%5D)
> > * [test_simple_add_new_node_while_query_info[1]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13271/testReport/junit/update_cluster_layout_tests/TestUpdateClusterLayout/Tests___dtest___test_simple_add_new_node_while_query_info_1_) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_simple_add_new_node_while_query_info%5B1%5D)
> > * [test_delete_tombstone_gc_node_down[SizeTieredCompactionStrategy]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13271/testReport/junit/compaction_test/TestCompaction/Tests___dtest___test_delete_tombstone_gc_node_down_SizeTieredCompactionStrategy_) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_delete_tombstone_gc_node_down%5BSizeTieredCompactionStrategy%5D)
> > * [test_delete_tombstone_gc_node_down[LeveledCompactionStrategy]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13271/testReport/junit/compaction_test/TestCompaction/Tests___dtest___test_delete_tombstone_gc_node_down_LeveledCompactionStrategy_) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_delete_tombstone_gc_node_down%5BLeveledCompactionStrategy%5D)
> > * [test_delete_tombstone_gc_node_down[TimeWindowCompactionStrategy]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13271/testReport/junit/compaction_test/TestCompaction/Tests___dtest___test_delete_tombstone_gc_node_down_TimeWindowCompactionStrategy_) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_delete_tombstone_gc_node_down%5BTimeWindowCompactionStrategy%5D)
> > * [test_delete_tombstone_gc_node_down[SizeTieredCompactionStrategy]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13271/testReport/junit/compaction_test/TestCompaction/Tests___dtest_with_topology_changes___test_delete_tombstone_gc_node_down_SizeTieredCompactionStrategy_) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_delete_tombstone_gc_node_down%5BSizeTieredCompactionStrategy%5D)
> > * [test_delete_tombstone_gc_node_down[LeveledCompactionStrategy]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13271/testReport/junit/compaction_test/TestCompaction/Tests___dtest_with_topology_changes___test_delete_tombstone_gc_node_down_LeveledCompactionStrategy_) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_delete_tombstone_gc_node_down%5BLeveledCompactionStrategy%5D)
> > * [test_delete_tombstone_gc_node_down[TimeWindowCompactionStrategy]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13271/testReport/junit/compaction_test/TestCompaction/Tests___dtest_with_topology_changes___test_delete_tombstone_gc_node_down_TimeWindowCompactionStrategy_) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_delete_tombstone_gc_node_down%5BTimeWindowCompactionStrategy%5D)
> > 
> > #### Build Details:
> > 
> > * Duration: 8 hr 43 min
> > * Builder: i-0797c39d6bf58af00 (m5ad.8xlarge)
> 
> @lkshminarayanan any idea on the failure?

I'm looking into this.

denesb commented:
> This is a bug fix and should be backported to older versions.

How old are we talking about? I see you added both 6.1 and 6.2, best to be specific when justifying the backport, with stating which commit introduced the regression or bug (or maybe it was like that since the start of times).

lkshminarayanan commented:
> > This is a bug fix and should be backported to older versions.
> 
> How old are we talking about? I see you added both 6.1 and 6.2, best to be specific when justifying the backport, with stating which commit introduced the regression or bug (or maybe it was like that since the start of times).

Ack. Updated the cover letter.

lkshminarayanan commented:
> > > #### 🔴 CI State: FAILURE
> > > ✅ - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13271/) ✅ - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13271/) **The following new/updated tests ran 100 times for each mode:** 🔹 boost/sstable_compaction_test ✅ - [Docker Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-docker/304/) ❌ - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/1799/) ❌ - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/12904/) ❌ - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/4892/) ✅ - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13271/)
> > > #### Failed Tests (10/36609):
> > > 
> > > * [test_delete_tombstone_gc_node_down[TimeWindowCompactionStrategy]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13271/testReport/junit/compaction_test/TestCompaction/Tests___dtest_with_tablets___test_delete_tombstone_gc_node_down_TimeWindowCompactionStrategy_) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_delete_tombstone_gc_node_down%5BTimeWindowCompactionStrategy%5D)
> > > * [test_delete_tombstone_gc_node_down[LeveledCompactionStrategy]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13271/testReport/junit/compaction_test/TestCompaction/Tests___dtest_with_tablets___test_delete_tombstone_gc_node_down_LeveledCompactionStrategy_) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_delete_tombstone_gc_node_down%5BLeveledCompactionStrategy%5D)
> > > * [test_delete_tombstone_gc_node_down[SizeTieredCompactionStrategy]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13271/testReport/junit/compaction_test/TestCompaction/Tests___dtest_with_tablets___test_delete_tombstone_gc_node_down_SizeTieredCompactionStrategy_) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_delete_tombstone_gc_node_down%5BSizeTieredCompactionStrategy%5D)
> > > * [test_simple_add_new_node_while_query_info[1]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13271/testReport/junit/update_cluster_layout_tests/TestUpdateClusterLayout/Tests___dtest___test_simple_add_new_node_while_query_info_1_) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_simple_add_new_node_while_query_info%5B1%5D)
> > > * [test_delete_tombstone_gc_node_down[SizeTieredCompactionStrategy]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13271/testReport/junit/compaction_test/TestCompaction/Tests___dtest___test_delete_tombstone_gc_node_down_SizeTieredCompactionStrategy_) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_delete_tombstone_gc_node_down%5BSizeTieredCompactionStrategy%5D)
> > > * [test_delete_tombstone_gc_node_down[LeveledCompactionStrategy]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13271/testReport/junit/compaction_test/TestCompaction/Tests___dtest___test_delete_tombstone_gc_node_down_LeveledCompactionStrategy_) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_delete_tombstone_gc_node_down%5BLeveledCompactionStrategy%5D)
> > > * [test_delete_tombstone_gc_node_down[TimeWindowCompactionStrategy]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13271/testReport/junit/compaction_test/TestCompaction/Tests___dtest___test_delete_tombstone_gc_node_down_TimeWindowCompactionStrategy_) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_delete_tombstone_gc_node_down%5BTimeWindowCompactionStrategy%5D)
> > > * [test_delete_tombstone_gc_node_down[SizeTieredCompactionStrategy]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13271/testReport/junit/compaction_test/TestCompaction/Tests___dtest_with_topology_changes___test_delete_tombstone_gc_node_down_SizeTieredCompactionStrategy_) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_delete_tombstone_gc_node_down%5BSizeTieredCompactionStrategy%5D)
> > > * [test_delete_tombstone_gc_node_down[LeveledCompactionStrategy]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13271/testReport/junit/compaction_test/TestCompaction/Tests___dtest_with_topology_changes___test_delete_tombstone_gc_node_down_LeveledCompactionStrategy_) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_delete_tombstone_gc_node_down%5BLeveledCompactionStrategy%5D)
> > > * [test_delete_tombstone_gc_node_down[TimeWindowCompactionStrategy]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13271/testReport/junit/compaction_test/TestCompaction/Tests___dtest_with_topology_changes___test_delete_tombstone_gc_node_down_TimeWindowCompactionStrategy_) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_delete_tombstone_gc_node_down%5BTimeWindowCompactionStrategy%5D)
> > > 
> > > #### Build Details:
> > > 
> > > * Duration: 8 hr 43 min
> > > * Builder: i-0797c39d6bf58af00 (m5ad.8xlarge)
> > 
> > 
> > @lkshminarayanan any idea on the failure?
> 
> I'm looking into this.

Changelog v4:

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13461/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13461/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/sstable_compaction_test
:white_check_mark: - [Docker Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-docker/422/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/1942/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/5050/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/13067/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13461/)
#### Build Details:
- Duration: 8 hr 39 min
- Builder: i-0eb2cf353354bd7a9 (m5ad.8xlarge)

lkshminarayanan commented:
@denesb @raphaelsc Hi! This is now ready for review.

raphaelsc reviewed (approved):
Looks good!

mykaul commented:
@scylladb/scylla-maint  - please review for merge.
